### PR TITLE
Introduce a namespace installer

### LINKED ISSUE #23932: Introduce ProcessNamespace?
In both https://github.com/servo/servo/pull/23637 and https://github.com/servo/servo/pull/23909 I find myself installing `PipelineNamspace` in various threads that currently do not have one, mainly as a way to use various new types of `Id` that aren't `PipelineId`, like `MessagePortId` or `IpcRouterId`. Those need to follow a similar model since they can end-up crossing process boundaries(message-port), or need to be unique per process. 

It can be quite hard to install a pipeline namespace for some threads, especially if those threads do not have a direct channel to the constellation. For example the threadpool used by layout, from within which calls are made to the font-cache(which I tried to refactor to use this prototype `SharedIpcRouter`). 

For other threads, like dedicated workers and layout, it's merely an inconvenient blocking ipc call to the constellation.

So, perhaps we could have a `ProcessNamespace` be a lazy static, put a mutex around it(or make it an atomic), then each thread upon initialization could grab it, install a thread-local `PipelineNamespace` based on something like `ProcessNamespace::new_pipeline_id()`, which could be an easy way for a thread to install a `PipelineNamespace` without having to communicate with the constellation. 

And the `ProcessNamespace`  is something the constellation could seed, like is currently done with `PipelineNamespace`, when the process is created.

### PR DESCRIPTION
**Pull request by gterzian**
---

- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #23932 (GitHub issue number if applicable)

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/23934)

### CODE REVIEW

**Comment by gterzian** [comment]
I think this is better than the current manual counting of pipeline-namespace that goes on in servo/lib and constellation, it becomes downright useful in the light of for example https://github.com/servo/servo/pull/23637

Where in order to install a pipeline-namespace in a worker thread, you need to make a blocking ipc call to the constellation, see https://github.com/servo/servo/pull/23637/files#diff-55c92a6a5ba7654ce45fe6fc6c63740fR1471

It would also be useful to install a pipeline-namespace in threads that do not have a direct channel to the constellation, for example the style threadpool(the reason why it might need one is anything requiring creating unique Ids, for example using the shared-ipc router prototyped at https://github.com/servo/servo/pull/23909, which currently can't work with parallel layout, as the communication with the font-cache occurs from a rayon thread from which it isn't easy to ask the constellation for a pipeline-namespace id).

**Comment by gterzian** [comment]
This isn't going to work in multiprocess, since there is nothing to prevent one process to race ahead of the other if the namespace is just a u32....

I think the only way is tot actually use the constellation as a global lock...

**Comment by gterzian** [comment]
Actually, the last commit might be something the could work. It means we'd still use the constellation to provide the `PipelineNamspaceId` to install a pipeline-namespace, and it would then be shared with all threads in the process, so each could use it to create Ids without having to install it's own pipeline. 

The downside is that the namespace becomes a kind of contended resource for all threads in a process, yet this might not matter in practice. If it were to become a problem, it could be solved by adding some intermediary per-thread namespace, that would be installed using the per-process namespace...

This could definitely make it much easier to start using the name-space from a bunch of other threads, without having to orchestrate installing it via the constellation.

**Comment by gterzian** [comment]
@jdm Ok here is the latest variation on that theme. I think this can be a way to address https://github.com/servo/servo/pull/23637/files#r312133197 generally across all threads. 

So, still using the constellation as a single source of unique per UA `PipelineNamespaceId`, and now adding a per-process `PipelineNameSpaceInstaller` that is shared with all threads in the process, and that can be used to setup thread-local pipeline-namespaces. 

It's better than the previous version since it keeps the actual namespace as a thread-local, and it improves on the current situation since it makes it easy for any thread to install a namespace. 

It improves on the code at https://github.com/servo/servo/pull/23637/files by removing the need for threads to have their own plumbing to ask the constellation for a namespace. For example, with this any thread in a rayon pool could get it's own namespace without needing to setup a channel with the constellation directly or otherwise being aware of how a namespace is installed. 

Still building to test if it actually works, I think it should be ok.

**Comment by gterzian** [comment]
@jdm this is is ready for review, I haven't squashed those commits since I'm not sure the "batching of Ids" is a good idea, so that commit could still be removed. 

This is somewhat relevant for the messageport PR, since it could rebase off of this one and simplify the worker code that installs pipeline namespaces.

File: components/constellation/constellation.rs [hunk 1]
```diff
@@ -818,6 +818,18 @@ where
         namespace_id
     }
 
+    /// Pre-generate a buffer of namespace ids, so content, or other, processes,
+    /// do not require to ipc for them initially.
```
Comment (jdm) [inline components/constellation/constellation.rs]: I imagine that any IPC for pipeline namespace generation will be completely dwarfed by the time it takes for the network request to complete. I don't think this is worth including without data backing it up.

**Review by jdm — CHANGES_REQUESTED** [review]
These changes look fine, but I don't think we need the batch generation.

**Review by jdm — APPROVED** [review]

**Comment by gterzian** [comment]
Ok, rebased as well...

**Pull request merged**